rules_python/piptool.py

```diff
@@ -93,6 +48,47 @@ def repository_name(self):
 parser.add_argument('--directory', action='store',
                     help=('The directory into which to put .whl files.'))
 
+def determine_possible_extras(whls):
+  whl_map = {
+    whl.distribution(): whl
+    for whl in whls
+  }
+  def parse_requirement(name):
+    if '[' not in name:
+      return name, None
+    assert name[-1] == ']'
+    (distribution, extra) = name[:-1].split('[')
+    return distribution, extra
+
+  # TODO(mattmoor): Consider memoizing if this recursion ever becomes
+  # expensive enough to warrant it.
+  def is_possible(distro, extra):
+    # If we don't have the .whl at all, then this isn't possible.
+    if distro not in whl_map:
+      return False
+    whl = whl_map[distro]
+    # If we have the .whl, and we don't need anything extra then
+    # we can satisfy this dependency.
+    if not extra:
+      return True
+    # If we do need something extra, then check the extra's
+    # dependencies to make sure they are fully satisfied.
+    for extra_dep in whl.dependencies(extra=extra):
+      dep_distro, dep_extra = parse_requirement(extra_dep)
+      if not is_possible(dep_distro, dep_extra):
```

Comment:
This doesn't work for me for with google-cloud-datastore 1.4.0.

The keys of `whl_map` have underscores (eg `"googleapis_common_protos"`), as does `distro`. `dep_distro` has dashes, eg `"googleapi-common-protos"`. So if an extra depends on something with dashes, then it fails to find it in `whl_map`, even though the version with underscores is there. For google-cloud-datastore, this occurs because eg proto-google-cloud-datastore-v1[grpc] depends on googleapis-common-protos[grpc].

See workaround: drigz/rules_python@2d6ce0bf